build-tools/src/main/java/org/elasticsearch/gradle/LoggedExec.java

```diff
@@ -7,81 +7,105 @@
+
+    @Input
+    abstract public Property<File> getWorkingDir();
+
+    private String output;
 
     @Inject
-    public LoggedExec(FileSystemOperations fileSystemOperations) {
+    public LoggedExec(ProjectLayout projectLayout, ExecOperations execOperations, FileSystemOperations fileSystemOperations) {
+        this.projectLayout = projectLayout;
+        this.execOperations = execOperations;
         this.fileSystemOperations = fileSystemOperations;
-        if (getLogger().isInfoEnabled() == false) {
-            setIgnoreExitValue(true);
-            setSpoolOutput(false);
-            // We use an anonymous inner class here because Gradle cannot properly snapshot this input for the purposes of
-            // incremental build if we use a lambda. This ensures LoggedExec tasks that declare output can be UP-TO-DATE.
-            doLast(new Action<Task>() {
-                @Override
-                public void execute(Task task) {
-                    int exitValue = LoggedExec.this.getExecutionResult().get().getExitValue();
-                    if (exitValue != 0) {
-                        try {
-                            LoggedExec.this.getLogger().error("Output for " + LoggedExec.this.getExecutable() + ":");
-                            outputLogger.accept(LoggedExec.this.getLogger());
-                        } catch (Exception e) {
-                            throw new GradleException("Failed to read exec output", e);
-                        }
-                        throw new GradleException(
-                            String.format(
-                                "Process '%s %s' finished with non-zero exit value %d",
-                                LoggedExec.this.getExecutable(),
-                                LoggedExec.this.getArgs(),
-                                exitValue
-                            )
-                        );
-                    }
-                }
-            });
-        }
+        getWorkingDir().convention(projectLayout.getProjectDirectory().getAsFile());
+        // For now mimic default behaviour of Gradle Exec task here
+        getEnvironment().putAll(System.getenv());
+        getCaptureOutput().convention(false);
     }
 
-    public void setSpoolOutput(boolean spoolOutput) {
-        final OutputStream out;
+    @TaskAction
+    public void run() {
+        Consumer<Logger> outputLogger;
+        OutputStream out = getCaptureOutput().get() ? new ByteArrayOutputStream() : System.out;
+        OutputStream effectiveOutputStream = out;
         if (spoolOutput) {
-            File spoolFile = new File(getProject().getBuildDir() + "/buffered-output/" + this.getName());
-            out = new LazyFileOutputStream(spoolFile);
+            File spoolFile = new File(projectLayout.getBuildDirectory().dir("buffered-output").get().getAsFile(), this.getName());
+            effectiveOutputStream = new TeeOutputStream(out, new LazyFileOutputStream(spoolFile));
```

Comment:
@mark-vieira I've reworked this with two implicit changes:
1. we now do not support capturing output now for later evaluation if spooling is enabled. I could be achieved by reading from the spool file but we currently have no use case for this
2. When indenting output is used (e.g. for our nested gradle builds) we do not log out to console by default anymore as this logger used to do so. This should be fine as it still logs to console in failure case. 
3. As an interesting side effect spooling the output for the nested gradle builds works correctly now. Before we configured spooling to be true, but then reset the output stream to the indenting output stream which delegated to console instead of using the earlier configured LazyOutputstream to the spooling file.

I added further integration tests for these scenarios and that the task failure case is handling correctly.

Reply:
> we now do not support capturing output now for later evaluation if spooling is enabled. I could be achieved by reading from the spool file but we currently have no use case for this

I think that's fine.

> When indenting output is used (e.g. for our nested gradle builds) we do not log out to console by default anymore as this logger used to do so. This should be fine as it still logs to console in failure case.

I'm on the fence about this. This is primiarily for building BWC distributions and these tasks can be long (over 1m). Having no console output makes it difficult to tell if something is hung or going wrong. My preference would be to keep the existing behavior of teeing to stdout in this scenario. Basically, the whole reason to indent to begin with is so you can differentiate the nested build from the outer one. If we're just spooling to a file and printing on error, the indenting wouldn't be necessary.

> As an interesting side effect spooling the output for the nested gradle builds works correctly now. Before we configured spooling to be true, but then reset the output stream to the indenting output stream which delegated to console instead of using the earlier configured LazyOutputstream to the spooling file.

It would seem to me that forwarding to stdout and spooling to disk are redundant. In a "normal" operation, we'd spool the output because we'd only display it in the case of a failure. If we are forwarding the output to the console, then why print it _again_ in the case of a failure.

It seems to me the BWC building case could just be done with a regular `Exec` task configured with an intending output stream. There's no need to otherwise "hold on to" the output and print it out later. `LoggedExec` should be used exclusively for shelling out where we don't need to see the output in the console, but we do want to keep it for reporting later.

Reply:
I've reworked the LoggedExec to fix the indention for nested builds logging to stdout. Using Exec is actually what we do not want anymore as we want to have those tasks configuration cache compatible. Anyhow we should disable spooling for those nested builds